File: shared/src/main/java/com/runtracker/shared/data/repository/NutritionRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.DailyNutritionDao
import com.runtracker.shared.data.db.NutritionGoalsDao
import com.runtracker.shared.data.model.*
import kotlinx.coroutines.flow.Flow
import java.util.*

class NutritionRepository(
    private val dailyNutritionDao: DailyNutritionDao,
    private val nutritionGoalsDao: NutritionGoalsDao
) {
    // Daily Nutrition
    fun getTodayNutrition(): Flow<DailyNutrition?> {
        return dailyNutritionDao.getDailyNutritionByDateFlow(getStartOfDay())
    }

    suspend fun getTodayNutritionOnce(): DailyNutrition? {
        return dailyNutritionDao.getDailyNutritionByDate(getStartOfDay())
    }

    fun getRecentNutrition(days: Int = 7): Flow<List<DailyNutrition>> {
        return dailyNutritionDao.getRecentNutritionFlow(days)
    }

    fun getNutritionInRange(startDate: Long, endDate: Long): Flow<List<DailyNutrition>> {
        return dailyNutritionDao.getNutritionInRangeFlow(startDate, endDate)
    }

    suspend fun getOrCreateTodayNutrition(
        userProfile: UserProfile?,
        goals: NutritionGoals?
    ): DailyNutrition {
        val today = getStartOfDay()
        val existing = dailyNutritionDao.getDailyNutritionByDate(today)
        
        if (existing != null) return existing

        // Calculate targets based on user profile and goals
        val effectiveGoals = goals ?: NutritionGoals()
        val targets = calculateDailyTargets(userProfile, effectiveGoals)

        val newNutrition = DailyNutrition(
            date = today,
            targetCalories = targets.calories,
            targetProteinGrams = targets.proteinGrams,
            targetCarbsGrams = targets.carbsGrams,
            targetFatGrams = targets.fatGrams,
            targetFiberGrams = targets.fiberGrams,
            targetWaterMl = userProfile?.weight?.let { 
                NutritionCalculator.calculateWaterTarget(it, effectiveGoals) 
            } ?: 2500
        )

        dailyNutritionDao.insertDailyNutrition(newNutrition)
        return newNutrition
    }

    suspend fun updateDailyNutrition(nutrition: DailyNutrition) {
        dailyNutritionDao.updateDailyNutrition(nutrition.copy(updatedAt = System.currentTimeMillis()))
    }

    suspend fun addMeal(meal: MealEntry) {
        val today = getTodayNutritionOnce() ?: return
        val updatedMeals = today.meals + meal
        val updatedNutrition = today.copy(
            meals = updatedMeals,
            consumedCalories = today.consumedCalories + meal.calories,
            consumedProteinGrams = today.consumedProteinGrams + meal.proteinGrams.toInt(),
            consumedCarbsGrams = today.consumedCarbsGrams + meal.carbsGrams.toInt(),
            consumedFatGrams = today.consumedFatGrams + meal.fatGrams.toInt(),
            consumedFiberGrams = today.consumedFiberGrams + meal.fiberGrams.toInt()
        )
        updateDailyNutrition(updatedNutrition)
    }

    suspend fun removeMeal(mealId: String) {
        val today = getTodayNutritionOnce() ?: return
        val meal = today.meals.find { it.id == mealId } ?: return
        val updatedMeals = today.meals.filter { it.id != mealId }
        val updatedNutrition = today.copy(
            meals = updatedMeals,
            consumedCalories = (today.consumedCalories - meal.calories).coerceAtLeast(0),
            consumedProteinGrams = (today.consumedProteinGrams - meal.proteinGrams.toInt()).coerceAtLeast(0),
            consumedCarbsGrams = (today.consumedCarbsGrams - meal.carbsGrams.toInt()).coerceAtLeast(0),
            consumedFatGrams = (today.consumedFatGrams - meal.fatGrams.toInt()).coerceAtLeast(0),
            consumedFiberGrams = (today.consumedFiberGrams - meal.fiberGrams.toInt()).coerceAtLeast(0)
        )
        updateDailyNutrition(updatedNutrition)
    }

    suspend fun addWater(ml: Int) {
        val today = getTodayNutritionOnce() ?: return
        val updatedNutrition = today.copy(waterMl = today.waterMl + ml)
        updateDailyNutrition(updatedNutrition)
    }

    suspend fun updateStepCount(steps: Int, weightKg: Double) {
        val today = getTodayNutritionOnce() ?: return
        val stepCalories = NutritionCalculator.calculateStepCalories(steps, weightKg)
        val updatedNutrition = today.copy(
            stepCount = steps,
            stepCaloriesBurned = stepCalories,
            burnedCalories = stepCalories + today.workoutCaloriesBurned
        )
        updateDailyNutrition(updatedNutrition)
    }

    suspend fun addWorkoutCalories(calories: Int) {
        val today = getTodayNutritionOnce() ?: return
        val updatedNutrition = today.copy(
            workoutCaloriesBurned = today.workoutCaloriesBurned + calories,
            burnedCalories = today.stepCaloriesBurned + today.workoutCaloriesBurned + calories
        )
        updateDailyNutrition(updatedNutrition)
    }
    
    suspend fun setWorkoutCalories(calories: Int) {
        val today = getTodayNutritionOnce() ?: return
        val updatedNutrition = today.copy(
            workoutCaloriesBurned = calories,
            burnedCalories = today.stepCaloriesBurned + calories
        )
        updateDailyNutrition(updatedNutrition)
    }

    // Nutrition Goals
    fun getGoals(): Flow<NutritionGoals?> {
        return nutritionGoalsDao.getGoalsFlow()
    }

    suspend fun getGoalsOnce(): NutritionGoals? {
        return nutritionGoalsDao.getGoals()
    }

    suspend fun saveGoals(goals: NutritionGoals) {
        nutritionGoalsDao.insertGoals(goals.copy(updatedAt = System.currentTimeMillis()))
    }

    // Stats
    suspend fun getWeeklyAverages(): NutritionWeeklyStats {
        val calendar = Calendar.getInstance().apply {
            add(Calendar.DAY_OF_YEAR, -7)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val weekAgo = calendar.timeInMillis
        val now = System.currentTimeMillis()

        return NutritionWeeklyStats(
            avgCalories = dailyNutritionDao.getAverageCaloriesInRange(weekAgo, now)?.toInt() ?: 0,
            avgProtein = dailyNutritionDao.getAverageProteinInRange(weekAgo, now)?.toInt() ?: 0,
            totalSteps = dailyNutritionDao.getTotalStepsInRange(weekAgo, now) ?: 0
        )
    }

    // Helpers
    private fun calculateDailyTargets(
        userProfile: UserProfile?,
        goals: NutritionGoals
    ): MacroTargets {
        val weight = userProfile?.weight ?: 70.0
        val height = userProfile?.height ?: 170.0
        val age = userProfile?.age ?: 30
        val gender = userProfile?.gender

        val calories = NutritionCalculator.calculateDailyCalories(
            weightKg = weight,
            heightCm = height,
            ageYears = age,
            gender = gender,
            activityLevel = goals.activityLevel,
            goal = goals.goal
        )

        return NutritionCalculator.calculateMacros(calories, weight, goals)
    }

    private fun getStartOfDay(): Long {
        return Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }
}

data class NutritionWeeklyStats(
    val avgCalories: Int,
    val avgProtein: Int,
    val totalSteps: Int
)

File: iosApp/GoSteady/Sources/Repositories/NutritionRepository.swift
import Foundation
import SwiftData

final class NutritionRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Daily Nutrition

    func fetchTodayNutrition() -> DailyNutrition? {
        let today = Date().startOfDay
        let descriptor = FetchDescriptor<DailyNutrition>(predicate: #Predicate { $0.date == today })
        return try? context.fetch(descriptor).first
    }

    func fetchNutritionForDate(_ date: Date) -> DailyNutrition? {
        let dayStart = date.startOfDay
        let descriptor = FetchDescriptor<DailyNutrition>(predicate: #Predicate { $0.date == dayStart })
        return try? context.fetch(descriptor).first
    }

    func fetchRecentNutrition(days: Int = 7) -> [DailyNutrition] {
        let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: Date())?.startOfDay ?? Date()
        let descriptor = FetchDescriptor<DailyNutrition>(
            predicate: #Predicate { $0.date >= cutoff },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchNutritionInRange(start: Date, end: Date) -> [DailyNutrition] {
        let s = start.startOfDay
        let e = end.startOfDay
        let descriptor = FetchDescriptor<DailyNutrition>(
            predicate: #Predicate { $0.date >= s && $0.date <= e },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func getOrCreateTodayNutrition(
        userWeight: Double?,
        userHeight: Double?,
        userAge: Int?,
        userGender: Gender?,
        goals: NutritionGoals?
    ) -> DailyNutrition {
        if let existing = fetchTodayNutrition() { return existing }

        let weight = userWeight ?? 70.0
        let height = userHeight ?? 170.0
        let age = userAge ?? 30
        let activityLevel = goals?.activityLevel ?? .moderatelyActive
        let goalType = goals?.goal ?? .maintain
        let proteinPerKg = goals?.proteinPerKgBodyweight ?? 1.6
        let carbPct = goals?.carbPercentage ?? 0.45
        let fatPct = goals?.fatPercentage ?? 0.25

        let calories = NutritionCalculator.calculateDailyCalories(
            weightKg: weight, heightCm: height, ageYears: age,
            gender: userGender, activityLevel: activityLevel, goal: goalType
        )
        let macros = NutritionCalculator.calculateMacros(
            totalCalories: calories, weightKg: weight,
            proteinPerKg: proteinPerKg, carbPercentage: carbPct, fatPercentage: fatPct
        )
        let waterTarget = NutritionCalculator.calculateWaterTarget(
            weightKg: weight, mlPerKg: goals?.waterPerKgBodyweight ?? 35.0
        )

        let nutrition = DailyNutrition(
            date: Date().startOfDay,
            targetCalories: macros.calories,
            targetProteinGrams: macros.proteinGrams,
            targetCarbsGrams: macros.carbsGrams,
            targetFatGrams: macros.fatGrams,
            targetFiberGrams: macros.fiberGrams,
            targetWaterMl: waterTarget
        )
        context.insert(nutrition)
        try? context.save()
        return nutrition
    }

    func updateNutrition(_ nutrition: DailyNutrition) {
        nutrition.updatedAt = Date()
        try? context.save()
    }

    // MARK: - Meal Management

    func addMeal(_ meal: MealEntry) {
        guard let today = fetchTodayNutrition() else { return }
        today.meals.append(meal)
        today.consumedCalories += meal.calories
        today.consumedProteinGrams += Int(meal.proteinGrams)
        today.consumedCarbsGrams += Int(meal.carbsGrams)
        today.consumedFatGrams += Int(meal.fatGrams)
        today.consumedFiberGrams += Int(meal.fiberGrams)
        updateNutrition(today)
    }

    func removeMeal(mealId: String) {
        guard let today = fetchTodayNutrition() else { return }
        guard let meal = today.meals.first(where: { $0.id == mealId }) else { return }
        today.meals.removeAll { $0.id == mealId }
        today.consumedCalories = max(0, today.consumedCalories - meal.calories)
        today.consumedProteinGrams = max(0, today.consumedProteinGrams - Int(meal.proteinGrams))
        today.consumedCarbsGrams = max(0, today.consumedCarbsGrams - Int(meal.carbsGrams))
        today.consumedFatGrams = max(0, today.consumedFatGrams - Int(meal.fatGrams))
        today.consumedFiberGrams = max(0, today.consumedFiberGrams - Int(meal.fiberGrams))
        updateNutrition(today)
    }

    func addWater(ml: Int) {
        guard let today = fetchTodayNutrition() else { return }
        today.waterMl += ml
        updateNutrition(today)
    }

    func updateStepCount(steps: Int, weightKg: Double) {
        guard let today = fetchTodayNutrition() else { return }
        let stepCalories = NutritionCalculator.calculateStepCalories(steps: steps, weightKg: weightKg)
        today.stepCount = steps
        today.stepCaloriesBurned = stepCalories
        today.burnedCalories = stepCalories + today.workoutCaloriesBurned
        updateNutrition(today)
    }

    func addWorkoutCalories(_ calories: Int) {
        guard let today = fetchTodayNutrition() else { return }
        today.workoutCaloriesBurned += calories
        today.burnedCalories = today.stepCaloriesBurned + today.workoutCaloriesBurned
        updateNutrition(today)
    }

    func setWorkoutCalories(_ calories: Int) {
        guard let today = fetchTodayNutrition() else { return }
        today.workoutCaloriesBurned = calories
        today.burnedCalories = today.stepCaloriesBurned + calories
        updateNutrition(today)
    }

    // MARK: - Nutrition Goals

    func fetchGoals() -> NutritionGoals? {
        let descriptor = FetchDescriptor<NutritionGoals>()
        return try? context.fetch(descriptor).first
    }

    func saveGoals(_ goals: NutritionGoals) {
        goals.updatedAt = Date()
        if fetchGoals() == nil {
            context.insert(goals)
        }
        try? context.save()
    }

    // MARK: - Stats

    func weeklyAverages() -> NutritionWeeklyStats {
        let recent = fetchRecentNutrition(days: 7)
        let count = max(1, recent.count)
        let avgCalories = recent.reduce(0) { $0 + $1.consumedCalories } / count
        let avgProtein = recent.reduce(0) { $0 + $1.consumedProteinGrams } / count
        let totalSteps = recent.reduce(0) { $0 + $1.stepCount }
        return NutritionWeeklyStats(avgCalories: avgCalories, avgProtein: avgProtein, totalSteps: totalSteps)
    }
}

struct NutritionWeeklyStats {
    let avgCalories: Int
    let avgProtein: Int
    let totalSteps: Int
}
